pkg/kallsyms/kallsyms_test.go

```diff
@@ -77,78 +82,151 @@ func TestRealKAllSyms(t *testing.T) {
+	t.Cleanup(resetState)
+
+	var err error
+	kAllSymsStr := strings.Join([]string{
+		"0000000000000000 A fixed_percpu_data",
+		"ffffffffb4231f40 D bpf_prog_fops",
+		"ffffffffb43723e0 d socket_file_ops",
+	},
+		"\n")
+	kAllSymsReader := strings.NewReader(kAllSymsStr)
+	kAllSyms, err := NewKAllSymsFromReader(kAllSymsReader)
+	require.Nil(t, err, "NewKAllSymsFromReader failed: %v", err)
+	kSymbolResolver := newKAllSymsResolver()
+	kSymbolResolver.kAllSyms = kAllSyms
+
+	// Little endian representation of the addresses above:
+	bpfProgFopsAddr := []byte{0x40, 0x1f, 0x23, 0xb4, 0xff, 0xff, 0xff, 0xff}
+	socketFileOpsAddr := []byte{0xe0, 0x23, 0x37, 0xb4, 0xff, 0xff, 0xff, 0xff}
+
+	spec := specTest.Copy()
+
+	err = specUpdateAddresses(
+		[]symbolResolver{
+			kSymbolResolver,
+		},
 		spec,
 		[]string{"abcde_bad_name"},
 	)
-	require.ErrorIs(t, err, os.ErrNotExist, "specRewriteConstantsWithSymbolAddresses should have failed")
+	require.ErrorContainsf(t, err, "was not found in kallsyms", "specUpdateAddresses should have failed")
 
 	err = specUpdateAddresses(
-		kAllSymsFactory,
+		[]symbolResolver{
+			kSymbolResolver,
+		},
 		spec,
 		[]string{"bpf_prog_fops", "socket_file_ops"},
 	)
-	require.Nil(t, err, "specRewriteConstantsWithSymbolAddresses failed: %v", err)
+	require.Nil(t, err, "specUpdateAddresses failed: %v", err)
 
 	expectedContents := []byte{}
 	expectedContents = append(expectedContents, bpfProgFopsAddr...)
 	expectedContents = append(expectedContents, socketFileOpsAddr...)
 	contents := spec.Maps[".rodata"].Contents[0].Value
 	require.Equal(t, contents, expectedContents, "contents aren't equal")
 }
+
+func TestSpecRewriteConstantsRoot(t *testing.T) {
+	utilstest.RequireRoot(t)
+	t.Cleanup(resetState)
+
+	resetState()
+
+	spec1 := specTest.Copy()
+	err := specUpdateAddresses(
+		[]symbolResolver{
+			newKAllSymsResolver(),
+		},
+		spec1,
+		[]string{"bpf_prog_fops", "socket_file_ops"},
+	)
+	require.Nil(t, err, "specUpdateAddresses failed: %v", err)
+
+	resetState()
+
+	spec2 := specTest.Copy()
+	err = specUpdateAddresses(
+		[]symbolResolver{
+			newKAllSymsResolver(),
+		},
+		spec2,
+		[]string{"bpf_prog_fops", "socket_file_ops"},
+	)
+	require.Nil(t, err, "specUpdateAddresses failed: %v", err)
+
+	resetState()
+
+	spec3 := specTest.Copy()
+	err = specUpdateAddresses(
+		[]symbolResolver{
+			newKAllSymsResolver(),
+			newEbpfResolver(),
+		},
+		spec3,
+		[]string{"bpf_prog_fops", "socket_file_ops"},
+	)
+	require.Nil(t, err, "specUpdateAddresses failed: %v", err)
+
+	resetState()
+
+	spec4 := specTest.Copy()
+	err = specUpdateAddresses(
+		[]symbolResolver{},
+		spec4,
+		[]string{"bpf_prog_fops", "socket_file_ops"},
+	)
+	require.ErrorIs(t, err, os.ErrNotExist, "specUpdateAddresses should have failed")
+
+	t.Logf("spec1: %v", spec1.Maps[".rodata"].Contents[0].Value)
+	t.Logf("spec2: %v", spec2.Maps[".rodata"].Contents[0].Value)
+	t.Logf("spec3: %v", spec3.Maps[".rodata"].Contents[0].Value)
```

Comment:
What's the idea behind your tests with spec1 and spec2? It seems to be the same test, and the cache is reset in between with `resetState()`.